### DISCUSSION TITLE
"type": "module" breaks .ts API endpoints

### DISCUSSION THREAD

**Discussion by MrAmericanMike** [vercel/community#1303] [Help]
### Summary

I have a simple project with just one api endpoint for testing under `api/random/index.ts` with this code:
```js
import type { VercelRequest, VercelResponse } from "@vercel/node";

function handler(req: VercelRequest, res: VercelResponse) {
	res.send(Math.floor(Math.random() * 100)).end();
}

export default handler;
```
The dev dependency installed is `"@vercel/node": "^2.8.14"`.

Doing `vercel dev` everything works as expected.

The problem comes when in `package.json` I add `"type": "module"` then the API endpoint breaks with this error:
```
(node:14364) ExperimentalWarning: Custom ESM Loaders is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
node:internal/errors:477
    ErrorCaptureStackTrace(err);
    ^

Error [ERR_UNSUPPORTED_ESM_URL_SCHEME]: Only URLs with a scheme in: file, data are supported by the default ESM loader. On Windows, absolute paths must be valid file:// URLs. Received protocol 'c:'       
    at new NodeError (node:internal/errors:387:5)
    at throwIfUnsupportedURLScheme (node:internal/modules/esm/resolve:1116:11)
    at nextResolve (node:internal/modules/esm/loader:165:28)
    at ESMLoader.resolve (node:internal/modules/esm/loader:844:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:431:18)
    at ESMLoader.import (node:internal/modules/esm/loader:528:22)
    at initializeLoader (node:internal/process/esm_loader:75:58)
    at loadESM (node:internal/process/esm_loader:90:11)
    at runMainESM (node:internal/modules/run_main:55:21) {
  code: 'ERR_UNSUPPORTED_ESM_URL_SCHEME'
}
Error: Function `api/random/index.ts` failed with exit code 1
```

I do need `"type": "module"` for Svelte to work.

What am I doing wrong?

**Comment by amyegan** [comment]
Thanks for reaching out @MrAmericanMike! It sounds like this might be related to a problem we faced earlier this month. Can you try the suggestions in https://github.com/vercel/community/discussions/1225#discussioncomment-4642381 and let me know if that solves the problem?

**Reply by MrAmericanMike** [reply]
> Thanks for reaching out @MrAmericanMike! It sounds like this might be related to a problem we faced earlier this month. Can you try the suggestions in [#1225 (comment)](https://github.com/orgs/vercel/discussions/1225#discussioncomment-4642381) and let me know if that solves the problem?

I tried, but still fails. I created a repository with the code I have in case I'm missing something https://github.com/MrAmericanMike/vercel-api-ts

PS: Tested with Node v16.17.1 and Node v18.13.0

Also, by having the tsconfig.json file with the suggested "esnext" causes index.ts to complain on the types import with: 
```
Cannot find module '@vercel/node'. Did you mean to set the 'moduleResolution' option to 'node', or to add aliases to the 'paths' option?
```

adding `"moduleResolution": "node"` fixes that.

**Reply by amyegan** [reply]
Thanks for sharing this repo! There might be an interaction happening between your code and your Project Settings. I'm not able to reproduce the error you're seeing. It even works for me without including "moduleResolution". 

Have you specified any [Build and Development Settings](https://vercel.com/docs/concepts/deployments/configure-a-build#build-and-development-settings) for the project? That might account for the difference in what you and I are seeing.

**Reply by MrAmericanMike** [reply]
I will test that, as it's a "bug" that I noticed also in some projects, the Build and Development Settings that exist on the deploy website can affect how the project works locally, maybe it's not a but and it's something by design, but I have noticed that for example, if a project was deployed with just and "api" and a "public" folder, if later on I add locally some framework to the project, then it will fail locally.

I didn't not set up anything. Maybe that's what I am missing?

![Screenshot 2023-01-25 170836](https://user-images.githubusercontent.com/8037559/214680336-cc475566-e938-4126-b07c-21aa9d4a0a12.png)

I'll clone the repository locally and deploy it as a new project to see what happens.

PS: It still fails - BUT WAIT!!! It fails locally, but not if I do a production deployment. In here it works https://vercel-api-ts-blond.vercel.app/api/random

![Screenshot 2023-01-25 171958](https://user-images.githubusercontent.com/8037559/214681165-b7a00236-e14c-432a-abee-0a895228b337.png)

I wonder what the issue could be. My main suspect is "pnpm" so I'll do some tests installing the packages with "npm" instead. After that, not sure what else I could test.

**Comment by MrAmericanMike** [comment]
Do you know if is there any repository with a working example of a package.json containing type module, while the serverless functions are written in Typescript that I should be able to clone to test? Thanks in advance.

PS: Installing the packages with npm made no difference.

**Comment by MrAmericanMike** [comment]
So I did more testing. Cloned the project into a Linux VM machine.

Ran with no issues using Vercel 23.1.2 each time I visited the endpoint locally it worked as expected.

So I downgraded to 23.1.2 on my Windows machine, that didn't make it work, but the error is now different.

```
Error! Failed to complete request to /api/random: Error: connect ECONNREFUSED ::1:65496
Error! Failed to complete request to /api/random: Error: connect ECONNREFUSED ::1:65520
Error! Failed to complete request to /api/random: Error: connect ECONNREFUSED ::1:65522
```
The last number changes on every request.

On the Linux machine (Node 16.15.1) after updating the Vercel cli to 28.13.1 the route still works, but the console outputs a warning.

```
(node:11142) ExperimentalWarning: --experimental-loader is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
```

Which is very similar to the one in Windows, with the difference that the Windows one just makes the function crash.

```
(node:15692) ExperimentalWarning: Custom ESM Loaders is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)     
node:internal/errors:491
    ErrorCaptureStackTrace(err);
    ^

Error [ERR_UNSUPPORTED_ESM_URL_SCHEME]: Only URLs with a scheme in: file, data are supported by the default ESM loader. On Windows, absolute paths must be valid file:// URLs. Received protocol 'c:'
    at new NodeError (node:internal/errors:400:5)
    at throwIfUnsupportedURLScheme (node:internal/modules/esm/resolve:1055:11)
    at defaultResolve (node:internal/modules/esm/resolve:1135:3)
    at nextResolve (node:internal/modules/esm/loader:163:28)
    at ESMLoader.resolve (node:internal/modules/esm/loader:842:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:424:18)
    at ESMLoader.import (node:internal/modules/esm/loader:525:22)
    at initializeLoader (node:internal/process/esm_loader:75:58)
    at loadESM (node:internal/process/esm_loader:90:11)
    at runMainESM (node:internal/modules/run_main:55:21) {
  code: 'ERR_UNSUPPORTED_ESM_URL_SCHEME'
}

Node.js v18.13.0
Error: Function `api/random/index.ts` failed with exit code 1
```

I'll see if I can test with Node 16.15.1 in Windows tomorrow, but don't think that will do much. At this point I'm pretty sure there is something that Node doesn't like (handles correctly) when running on Windows, independent of Node version.

PS: Testing in 16.15.1 and no difference, it still fails.

**Comment by the1812** [comment]
Node.js has a [strange behavior](https://github.com/nodejs/node/issues/31710), which requires using `pathToFileURL` to process absolute path.

The path of the Function file is correctly processed.
https://github.com/vercel/vercel/blob/8f499695853c6689cb231c6f3f6020feb95b4713/packages/node-bridge/launcher.js#L69-L71

The ESM loader itself is also an absolute path **but not processed**. This crashes your Function.
https://github.com/vercel/vercel/blob/8f499695853c6689cb231c6f3f6020feb95b4713/packages/node/src/index.ts#L546-L547
https://github.com/vercel/vercel/blob/8f499695853c6689cb231c6f3f6020feb95b4713/packages/node/src/index.ts#L611-L616

Example fix using pnpm patch: https://github.com/the1812/Doujin-Meta/commit/5e9f0f5822f3f7485c481b527d144aaaf6498dcc
